File: .agents/skills/weaver-oa-dev/scripts/table_search.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
泛微OA (Ecology) 1,700 张核心数据库表快速检索工具 (Python 版)
"""

import sys
import os
import json
import argparse
import io

if sys.platform == 'win32':
    sys.stdout = io.TextIOWrapper(sys.stdout.buffer, encoding='utf-8')

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
DB_DICT_PATH = os.path.join(SCRIPT_DIR, "db_dictionary.json")

def load_dict():
    if os.path.exists(DB_DICT_PATH):
        with open(DB_DICT_PATH, 'r', encoding='utf-8') as f:
            return json.load(f)
    print(f"错误: 找不到表字典索引文件 {DB_DICT_PATH}")
    return []

def search_tables(keyword=None, module_name=None, limit=20, show_detail=False):
    tables = load_dict()
    kw = keyword.lower().strip() if keyword else None
    mod = module_name.lower().strip() if module_name else None

    results = []

    for t in tables:
        tbl_name = t.get('tableName', '')
        m_name = t.get('module', '')
        cols = t.get('columns', [])

        if mod and mod not in m_name.lower():
            continue

        is_match = False
        matched_col = None

        if not kw:
            is_match = True
        elif kw in tbl_name.lower() or kw in m_name.lower():
            is_match = True
        else:
            for c in cols:
                c_name = c.get('name', '').lower()
                c_cn = c.get('cn', '').lower()
                c_memo = c.get('memo', '').lower()
                if kw in c_name or kw in c_cn or kw in c_memo:
                    is_match = True
                    matched_col = c
                    break

        if is_match:
            t_copy = dict(t)
            t_copy['matchedColumn'] = matched_col
            results.append(t_copy)

    print(f"\n共检索到 {len(results)} 张匹配数据表 (展示前 {min(limit, len(results))} 个):\n")
    print(f"{'序号':<4} | {'所属模块':<12} | {'数据库表名':<32} | 字段数 | 字段预览 / 匹配字段")
    print("-" * 110)

    for idx, r in enumerate(results[:limit], 1):
        preview = ", ".join([f"{c.get('name')}({c.get('cn', '-')})" for c in r.get('columns', [])[:3]])
        if r.get('matchedColumn'):
            mc = r['matchedColumn']
            preview = f"[匹配列: {mc.get('name')}({mc.get('cn')})] " + preview
        print(f"{idx:<4} | {r.get('module', ''):<12} | {r.get('tableName', ''):<32} | {str(r.get('columnCount', 0)):<6} | {preview}")

    print("-" * 110)

    if len(results) == 1 or show_detail:
        target = results[0]
        print(f"\n=== 数据表 [{target.get('tableName')}] 完整字段结构 ({target.get('module')}) ===")
        print("| 序号 | 列名 (Column) | 中文说明 | 数据类型 | 长度 | 允许空 | 备注 |")
        print("| :---: | :--- | :--- | :---: | :---: | :---: | :--- |")
        for idx, c in enumerate(target.get('columns', []), 1):
            print(f"| {idx:<3} | {c.get('name', '-'):<20} | {c.get('cn', '-'):<16} | {c.get('type', '-'):<10} | {str(c.get('len', '-')):<6} | {c.get('nullable', '-'):<4} | {c.get('memo', '-')} |")
        print("")
    elif len(results) > limit:
        print(f"提示: 还有 {len(results) - limit} 张表未展示，可通过指定具体表名查看完整字段定义。\n")

def main():
    parser = argparse.ArgumentParser(description="泛微OA 数据库表快速检索工具")
    parser.add_argument("keyword", nargs="?", default=None, help="搜索关键词 (表名、字段名、中文名)")
    parser.add_argument("-m", "--module", default=None, help="按模块过滤 (如: 工作流程, 人力资源)")
    parser.add_argument("-n", "--limit", type=int, default=20, help="限制展示条数 (默认20)")
    parser.add_argument("-d", "--detail", action="store_true", help="显示第一条结果的完整字段列表")

    args = parser.parse_args()

    if not args.keyword and not args.module:
        parser.print_help()
        return

    search_tables(args.keyword, args.module, args.limit, args.detail)

if __name__ == '__main__':
    main()

File: skills/weaver-oa-dev/scripts/table_search.js
#!/usr/bin/env node
/**
 * 泛微OA (Ecology) 1,700 张核心数据库表快速检索工具 (Node.js 版)
 * 支持按表名（如 workflow_requestbase）、字段名（如 requestname）、模块名或中文描述检索
 * 运行方式:
 *   node scripts/table_search.js workflow_currentoperator
 *   node scripts/table_search.js 待办
 *   node scripts/table_search.js -m 人力资源
 */

const fs = require('fs');
const path = require('path');

const DB_DICT_PATH = path.join(__dirname, 'db_dictionary.json');

function loadDict() {
  if (fs.existsSync(DB_DICT_PATH)) {
    return JSON.parse(fs.readFileSync(DB_DICT_PATH, 'utf8'));
  }
  console.error('错误: 找不到表字典索引文件 ' + DB_DICT_PATH);
  return [];
}

function searchTables({ keyword = null, moduleName = null, limit = 20, showDetail = false }) {
  const tables = loadDict();
  const kw = keyword ? keyword.toLowerCase().trim() : null;
  const mod = moduleName ? moduleName.toLowerCase().trim() : null;

  const results = [];

  for (const t of tables) {
    const tableName = t.tableName || '';
    const module = t.module || '';
    const cols = t.columns || [];

    if (mod && !module.toLowerCase().includes(mod)) {
      continue;
    }

    let matchedColumn = null;
    let isMatch = false;

    if (!kw) {
      isMatch = true;
    } else if (tableName.toLowerCase().includes(kw) || module.toLowerCase().includes(kw)) {
      isMatch = true;
    } else {
      // Search inside columns
      for (const c of cols) {
        if (
          (c.name && c.name.toLowerCase().includes(kw)) ||
          (c.cn && c.cn.toLowerCase().includes(kw)) ||
          (c.memo && c.memo.toLowerCase().includes(kw))
        ) {
          isMatch = true;
          matchedColumn = c;
          break;
        }
      }
    }

    if (isMatch) {
      results.push({ ...t, matchedColumn });
    }
  }

  console.log(`\n共检索到 ${results.length} 张匹配数据表 (展示前 ${Math.min(limit, results.length)} 个):\n`);
  console.log(`${'序号'.padEnd(4)} | ${'所属模块'.padEnd(12)} | ${'数据库表名'.padEnd(32)} | 字段数 | 字段预览 / 匹配字段`);
  console.log('-'.repeat(110));

  results.slice(0, limit).forEach((r, idx) => {
    const indexStr = String(idx + 1).padEnd(4);
    const modStr = r.module.padEnd(12);
    const tableStr = r.tableName.padEnd(32);
    const countStr = String(r.columnCount).padEnd(6);
    let preview = r.columns
      .slice(0, 3)
      .map((c) => `${c.name}(${c.cn || '-'})`)
      .join(', ');
    if (r.matchedColumn) {
      preview = `[匹配列: ${r.matchedColumn.name}(${r.matchedColumn.cn})] ` + preview;
    }
    console.log(`${indexStr} | ${modStr} | ${tableStr} | ${countStr} | ${preview}`);
  });

  console.log('-'.repeat(110));

  // If user searched for an exact single table or requested detail, print its full schema
  if (results.length === 1 || showDetail) {
    const target = results[0];
    console.log(`\n=== 数据表 [${target.tableName}] 完整字段结构 (${target.module}) ===`);
    console.log(`| 序号 | 列名 (Column) | 中文说明 | 数据类型 | 长度 | 允许空 | 备注 |`);
    console.log(`| :---: | :--- | :--- | :---: | :---: | :---: | :--- |`);
    target.columns.forEach((c, idx) => {
      console.log(
        `| ${String(idx + 1).padEnd(3)} | ${(c.name || '-').padEnd(20)} | ${(c.cn || '-').padEnd(16)} | ${(c.type || '-').padEnd(10)} | ${(c.len || '-').padEnd(6)} | ${(c.nullable || '-').padEnd(4)} | ${c.memo || '-'} |`
      );
    });
    console.log('');
  } else if (results.length > limit) {
    console.log(`提示: 还有 ${results.length - limit} 张表未展示，可通过指定具体表名查看完整字段定义。\n`);
  }
}

function main() {
  const args = process.argv.slice(2);
  let keyword = null;
  let moduleName = null;
  let limit = 20;
  let showDetail = false;

  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    if (arg === '-m' || arg === '--module') {
      moduleName = args[++i];
    } else if (arg === '-n' || arg === '--limit') {
      limit = parseInt(args[++i], 10) || 20;
    } else if (arg === '-d' || arg === '--detail') {
      showDetail = true;
    } else if (!arg.startsWith('-')) {
      keyword = arg;
    }
  }

  if (!keyword && !moduleName) {
    console.log('泛微OA 数据库表快速检索工具 (Node.js 版)');
    console.log('用法:');
    console.log('  node scripts/table_search.js <表名或中文关键词>');
    console.log('  node scripts/table_search.js -m <模块名>');
    console.log('\n示例:');
    console.log('  node scripts/table_search.js workflow_currentoperator');
    console.log('  node scripts/table_search.js 考勤');
    console.log('  node scripts/table_search.js 审批意见 -d');
    return;
  }

  searchTables({ keyword, moduleName, limit, showDetail });
}

main();
